File: backend/src/payroll_copilot/application/dto/analytics.py
"""Application DTOs for analytics query results.

These are the Stable contract shapes future dashboards consume.
New metrics should add new DTOs rather than mutating existing ones.
"""

from __future__ import annotations

from dataclasses import dataclass, field
from decimal import Decimal
from typing import Any
from uuid import UUID


@dataclass(frozen=True, slots=True)
class PeriodPoint:
    """One payroll period bucket (period_year + period_month only)."""

    period_year: int
    period_month: int

    @property
    def label(self) -> str:
        return f"{self.period_year:04d}-{self.period_month:02d}"


@dataclass(frozen=True, slots=True)
class SalaryMonthPoint:
    period_year: int
    period_month: int
    net_salary: Decimal | None = None
    gross_salary: Decimal | None = None
    currency: str = "ILS"
    document_id: UUID | None = None
    extraction_id: UUID | None = None


@dataclass(frozen=True, slots=True)
class EmployeeSalaryAnalytics:
    employee_id: UUID
    organization_id: UUID
    year: int
    months: list[SalaryMonthPoint] = field(default_factory=list)
    available_years: list[int] = field(default_factory=list)
    documents_missing_period: int = 0


@dataclass(frozen=True, slots=True)
class OutcomeMonthPoint:
    period_year: int
    period_month: int
    documents_processed: int = 0
    success: int = 0
    review_required: int = 0
    failed: int = 0


@dataclass(frozen=True, slots=True)
class ValidationFailureMonthPoint:
    period_year: int
    period_month: int
    failure_count: int = 0
    runs_with_failures: int = 0


@dataclass(frozen=True, slots=True)
class ErrorTypeBucket:
    key: str
    count: int
    category: str | None = None


@dataclass(frozen=True, slots=True)
class ConfidenceMonthPoint:
    period_year: int
    period_month: int
    average_confidence: float | None = None
    sample_count: int = 0


@dataclass(frozen=True, slots=True)
class OrgPayrollAnalytics:
    organization_id: UUID
    year: int
    documents_by_month: list[OutcomeMonthPoint] = field(default_factory=list)
    validation_failures_by_month: list[ValidationFailureMonthPoint] = field(default_factory=list)
    error_type_distribution: list[ErrorTypeBucket] = field(default_factory=list)
    top_validation_failures: list[ErrorTypeBucket] = field(default_factory=list)
    average_confidence_by_month: list[ConfidenceMonthPoint] = field(default_factory=list)
    documents_missing_period: int = 0
    available_years: list[int] = field(default_factory=list)


@dataclass(frozen=True, slots=True)
class AccountantCaseload:
    payroll_accountant_id: UUID
    employee_count: int


@dataclass(frozen=True, slots=True)
class OrganizationCensusSlice:
    organization_id: UUID
    employees_count: int = 0
    payroll_accountants_count: int = 0
    employees_without_payroll_accountant: int = 0
    employees_per_payroll_accountant: list[AccountantCaseload] = field(default_factory=list)


@dataclass(frozen=True, slots=True)
class AdminOrgCensus:
    companies_count: int
    employees_count: int
    payroll_accountants_count: int
    employees_without_payroll_accountant: int
    employees_per_payroll_accountant: list[AccountantCaseload] = field(default_factory=list)
    organizations: list[OrganizationCensusSlice] = field(default_factory=list)
    extras: dict[str, Any] = field(default_factory=dict)

File: frontend/src/types/analytics.ts
/** Analytics API response types (Phase 1A backend contracts). */

export type SalaryMonthPoint = {
  period_year: number;
  period_month: number;
  net_salary: number | string | null;
  gross_salary: number | string | null;
  currency: string;
  document_id: string | null;
  extraction_id: string | null;
};

export type EmployeeSalaryAnalytics = {
  employee_id: string;
  organization_id: string;
  year: number;
  months: SalaryMonthPoint[];
  available_years: number[];
  documents_missing_period: number;
};

export type OutcomeMonthPoint = {
  period_year: number;
  period_month: number;
  documents_processed: number;
  success: number;
  review_required: number;
  failed: number;
};

export type ValidationFailureMonthPoint = {
  period_year: number;
  period_month: number;
  failure_count: number;
  runs_with_failures: number;
};

export type ErrorTypeBucket = {
  key: string;
  count: number;
  category: string | null;
};

export type ConfidenceMonthPoint = {
  period_year: number;
  period_month: number;
  average_confidence: number | null;
  sample_count: number;
};

export type OrgPayrollAnalytics = {
  organization_id: string;
  year: number;
  documents_by_month: OutcomeMonthPoint[];
  validation_failures_by_month: ValidationFailureMonthPoint[];
  error_type_distribution: ErrorTypeBucket[];
  top_validation_failures: ErrorTypeBucket[];
  average_confidence_by_month: ConfidenceMonthPoint[];
  documents_missing_period: number;
  available_years: number[];
};

export type AccountantCaseload = {
  payroll_accountant_id: string;
  employee_count: number;
};

export type OrganizationCensusSlice = {
  organization_id: string;
  employees_count: number;
  payroll_accountants_count: number;
  employees_without_payroll_accountant: number;
  employees_per_payroll_accountant: AccountantCaseload[];
};

export type AdminOrgCensus = {
  companies_count: number;
  employees_count: number;
  payroll_accountants_count: number;
  employees_without_payroll_accountant: number;
  employees_per_payroll_accountant: AccountantCaseload[];
  organizations: OrganizationCensusSlice[];
};

export type ConfidenceBucket = {
  label: string;
  min_inclusive: number;
  max_exclusive: number;
  count: number;
};

export type QualityMonthPoint = {
  period_year: number;
  period_month: number;
  documents_processed: number;
  extraction_attempted: number;
  extraction_success: number;
  extraction_success_rate: number | null;
  ocr_attempted: number;
  ocr_success: number;
  ocr_failed: number;
  validation_runs: number;
  validation_pass: number;
  validation_success_rate: number | null;
  average_confidence: number | null;
  confidence_sample_count: number;
  manual_review: number;
  manual_review_rate: number | null;
  failed_documents: number;
};

export type OrgQualityAnalytics = {
  organization_id: string;
  year: number;
  months: QualityMonthPoint[];
  confidence_distribution: ConfidenceBucket[];
  totals: QualityMonthPoint | null;
  documents_missing_period: number;
  available_years: number[];
};

export type AdminQualityAnalytics = {
  year: number;
  organizations_count: number;
  months: QualityMonthPoint[];
  confidence_distribution: ConfidenceBucket[];
  totals: QualityMonthPoint | null;
  organizations: OrgQualityAnalytics[];
  available_years: number[];
};
